### PR TITLE
fused_mla: contiguous rope tiles, out-of-place Q rotation, deterministic launch config

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Human Note:

with 4k token, inplace version is 14.1us, out-of-place version is 60.5us. 
assuming 60layer x 32 accumulate step, the step time difference is +89.1 ms. 
so 0.8% of the e2e time. 

Not small. 

but I can't think of a general solution inplace version. It either need to hack the SAC to recompute q_projection. 
or need the q_projection mxfp8 linear to own its output. 

this introduce the unwanted coupling between an kernel implementation and other parts of the system.

## Summary

Three related changes to the opt-in DeepSeek-V3 fused MLA override. Each is
independent and could be split if reviewers prefer.

1. **Load and store each head's RoPE tail as one contiguous tile.** All three
   kernels addressed the adjacent complex pairs as two stride-2 accesses over the
   same bytes, which prevents vectorization.
2. **Rotate Q out of place.** Rotating a view of the query projection in place
   forced `ctx.mark_dirty`, which cost a full-size `CopySlices` in backward.
3. **Autotune the launch config, pinned to a preset under determinism.** The
   right head tile depends on head count, but a benchmarked pick is not
   reproducible, and for these kernels it is not even numerically stable.

Also adds `enable_fused_mla()` next to the existing `enable_fused_swiglu()`.

Measurements are single-GPU GB300, `torch 2.15.0a0` / Triton 3.8, CUDA-graph
timed so they exclude `torch.library` dispatch overhead (~25-40us per call,
enough to hide a 2x difference on kernels this size).

---

## 1. Contiguous rope tiles

The kernels loaded `t[..., 0::2]` and `t[..., 1::2]` separately. Each head's
rope tail is 64 contiguous bf16 elements -- one 128-byte cache line -- so the
stride-2 form splits it into two half-empty gathers. The tile form loads it
whole and de-interleaves in registers with `tl.split` / `tl.join`.

The generated PTX confirms the intent, per program:

| | loads | stores |
| --- | --- | --- |
| before | 32x `ld.global.b16` | 32x `st.global.b16` |
| after | 4x `ld.global.v4.b32` | 4x `st.global.v4.b32` |

At the 671B shape (`B=1`, 4096 tokens, 128 heads):

| kernel | before | after | |
| --- | ---: | ---: | ---: |
| `q_rope` | 26.7us | 14.3us | 1.87x |
| `k_rope` | 52.4us | 52.1us | 1.01x |
| `kv_backward` | 93.2us | 90.6us | 1.03x |
| total | 172.4us | **157.0us** | 1.10x |

K and KV-backward do not move because they are already bandwidth bound at
~6.4 TB/s on their contiguous nope/value copies; only about a fifth of their
traffic is the rope tail.

**Note for anyone comparing against the internal writeup of this change:** it
reported 3.54x aggregate on Triton 3.6. On Triton 3.8 the pre-fix baseline is
already 3-8x faster than that report's, so most of the win appears to have been
absorbed upstream. 1.10x is what this actually buys today.

## 2. Out-of-place Q rotation

`Attention.forward` builds `q = proj.view(N, H, D).unsqueeze(0)` and handed that
view to the rotation. Mutating a view of a tensor with autograd history makes
autograd insert `CopySlices`, whose backward allocates a buffer shaped like the
**base**, not the mutated region. Every `aten` op in backward that allocates or
copies, at the 671B Q shape (201,326,592 B):

```
mutate through a view
  aten.copy_.default              (4096, 128, 192)      201,326,592 B
  aten.new_empty_strided.default  (4096, 24576)         201,326,592 B
  aten.copy_.default              (4096, 24576)         201,326,592 B
  aten.clone.default              (1, 4096, 128, 192)   201,326,592 B
  aten.copy_.default              (1, 4096, 128, 192)   201,326,592 B
  TOTAL                                               1,006,632,960 B

mutate an owned tensor
  (none)
```

Roughly 960 MiB of traffic to support one 192 MiB tensor. Because the view
covers 100% of the base there is no region outside it, so the base-sized seeding
copy is entirely overwritten -- PyTorch does not special-case coverage.

The fix is to not mutate at all. `_fused_q_rope_kernel` gains an output
pointer and a `COPY_NOPE` flag. The in-place op passes `q` as both source and
destination, so it loads and stores identical addresses and its output is
unchanged; the two variants share one kernel body rather than duplicating it.
The functional op allocates and lets the kernel carry the untouched nope half
across in the same pass.

Fusing that copy is the point. Three ways to produce the non-rotated half:

| | time |
| --- | ---: |
| `out[..., :nope] = q[..., :nope]`, then rotate the tail | 161.4us |
| `out = q.clone()`, then rotate in place | 75.7us |
| **kernel carries the nope half in the rotation pass** | **61.2us** |

The strided partial copy moves the fewest bytes and is the slowest, at ~1.9 TB/s
against 6.6 TB/s for the fused version -- copying 2/3 of each row is worse than
copying all of it contiguously.

Net effect: forward gets about 46us more expensive, backward loses ~1 GB of
copies. Q path forward+backward, in the view-shaped layout `Attention.forward`
actually produces: **5698 -> 5500 us/iter**.

## 3. Autotuned launch config, pinned under determinism

The best `(BLOCK_H, num_warps)` tracks head count: at 128 heads any tile from 16
up is within a few percent, but at 16 heads a single warp beats four by 1.18x on
the KV backward, which no single fixed default covers.

Autotuning alone would not be safe here. The pick is noise-sensitive -- six
identical runs chose `(32,1)` three times and `(128,4)` three times -- and those
two configs produce **different output**: five distinct results across eight
configs at the production shape. The cause is that these kernels compute
`q_even*cos - q_odd*sin`, and which of the two products is folded into the FMA
rather than rounded on its own depends on the tile. Both configs emit the same
op mix (2 `mul` : 2 `fma` : 1 `neg` per output pair), so it is not
FMA-vs-no-FMA; the association differs. Wherever the two products cancel -- the
differing values are ~7e-06 from operands of order 1 -- that is one bf16 ulp.

So under `torch.are_deterministic_algorithms_enabled()`, an
`early_config_prune` returns a single config, which makes Triton skip
benchmarking outright (`autotuner.py`: `if len(pruned_configs) == 1`). Otherwise
the full list is returned and autotuning proceeds. This mirrors what
`torchtitan/distributed/utils.py` already does for FlexAttention, disabling its
autotuning under determinism for the same reason.

That predicate is exactly `debug.deterministic`: `set_determinism()` keeps no
state of its own, and the flag reads `True` under `deterministic_warn_only` too.

Presets are the fastest measured at 128 heads and hold up as tokens scale:

| kernel | preset | 4k tok | 32k tok | 64k tok |
| --- | --- | ---: | ---: | ---: |
| Q rope | `128, 4` | +0.0% | +0.4% | +0.7% |
| K rope | `32, 4` | +0.2% | +0.4% | +0.5% |
| KV backward | `128, 4` | +0.0% | +0.0% | +0.0% |

(percentages are vs the per-shape optimum). One config serves both Q variants,
so `128,4` minimizes their sum: the functional forward is nearly flat across
tiles while the in-place backward is not, so the backward decides.

The one deliberate trade is the KV backward, where `128,4` is fastest at 128
heads but a single warp would be ~18% faster at 16.

`COPY_NOPE` is deliberately **not** in the autotune key, so both Q variants share
one tuning result and stay bitwise equal to each other.

---

## Numerics

- The in-place path is **bitwise identical** to the pre-change kernels on all
  five outputs (Q forward, Q inverse, K, `grad_kv`, `grad_k_pe`), across
  `B=1/2/4/8` and 4k-64k tokens, in bf16 and fp32.
- The out-of-place forward produces gradients **bitwise identical to stock
  (non-fused) attention** -- max abs diff `0.000e+00` on the debugmodel.
- `CopySlices` is gone from the real `Attention` autograd graph.
- Under determinism the config and output hash are stable across fresh
  processes; without it they vary, by design.

I did not run an end-to-end training loss comparison. The kernel-level bitwise
results above are stronger evidence for parts 1 and 2, but part 3 does mean a
**non-deterministic run is not bitwise comparable to another** -- which is why
determinism pins the config, and matches how FlexAttention autotuning is already
handled.

## Test plan

`pytest tests/unit_tests/cpu/test_fused_mla_override.py` -- 17 pass. New
coverage:

- functional and in-place Q variants agree bitwise, both rotation directions
- the functional variant leaves its input untouched
- determinism yields exactly one config; non-determinism yields the full list
- `COPY_NOPE` stays out of the autotune key
- `restore_value=["q"]` is set (the autotuner must restore between trials, or
  every trial past the first measures doubly rotated data)
- `make_fx` sees one functional op in forward and one in-place op in backward
- `enable_fused_mla()` appends once and asserts on double-enable

`pre-commit run --files <changed>` clean.

### CODE REVIEW

**Comment by SherlockNoMad** [comment]
cc @sanketpurandare 

2 difference from what you did in the diff

1. didn't introduce a config for block/wrap selection. using triton.autotune instead.
2. didn't include the SAC workaround. I am just doing the Q rotation out of place, it's costs a bit of perf, but is a more general solution, that doesn't need SAC special treatment, and doensn't need mxfp8 linear to own its output buffer.oensn't need mxfp8 linear to own its output buffer.

File: torchtitan/models/deepseek_v3/config_registry.py [hunk 1]
```diff
@@ -36,6 +36,14 @@ def enable_fused_swiglu(config: Trainer.Config) -> None:
         config.override.imports.append(override)
 
 
+def enable_fused_mla(config: Trainer.Config) -> None:
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/config_registry.py:39]: is this used anywhere?

File: torchtitan/overrides/fused_mla.py
Comment (tianyu-l) [inline torchtitan/overrides/fused_mla.py:1]: regarding

> It either need to hack the SAC to recompute q_projection. or need the q_projection mxfp8 linear to own its output.

Would it be doable with torch_remat, which we'll migrate over to soon.

File: torchtitan/models/deepseek_v3/config_registry.py (hunk 1 above)
Comment (SherlockNoMad) [reply torchtitan/models/deepseek_v3/config_registry.py]: no... removed.

I was following patterns of "enable_fused_swiglu", but it seems removed recently.

File: torchtitan/overrides/fused_mla.py
Comment (SherlockNoMad) [reply torchtitan/overrides/fused_mla.py:1]: No, this is not a problem of SAC or torch_remat per se. It's a problem of autograd engine would automatically insert a copy node for **a view of the activation buffers that has been inplace modified**, aka marked_dirty. 

In the original impl, q_projection is inplace rotated for that last 64 elements, and it's "marked_dirty" in the autograd.Function's forward. 

In the new impl, I am doing q_projection rotation out-of-place. so first 128 elements is copied over, and last 64 element is rotated. This has the performance cost. 

with 4k token, inplace version is 14.1us, out-of-place version is 60.5us.
assuming 60layer x 32 accumulate step, the step time difference is +89.1 ms.
so 0.8% of the e2e time.

Not small.

Alternatively, I can give up on the generality of this kernel, co-design the model such that it doesn't do the view of the inplace-modified buffer, so autograd engine doesn't introduce the addition copy.

Comment (anijain2305) [reply torchtitan/overrides/fused_mla.py:1]: @sanketpurandare is fixing the mxfp8 linear to own the outputs - https://github.com/pytorch/torchtitan/pull/4665

I think changing mxfp8 linear is easy, and therefore its ok to have that complexity in the mxfp8 linear.

Does that change this entire PR?

Comment (sanketpurandare) [reply torchtitan/overrides/fused_mla.py:1]: yeah mxfp8 linear can own the output when a config is enabled. So when fused MLA override is used it enables that automatically in mxfp8 linear.

Comment (anijain2305) [reply torchtitan/overrides/fused_mla.py:1]: Just sharing a one file repro to dive deeper - https://www.internalfb.com/phabricator/paste/view/P2516170608

(Will do that tomorrow)

Comment (anijain2305) [reply torchtitan/overrides/fused_mla.py:1]: **For simplicity, I would choose Q RoPE implementation.**

Background
------------
This is the flow

--> MxFP8 Linear **autograd.Function** returns a view output 
--> attention view/unsqueeze ops
--> Fused MLA Q Rope

The question is whether fused Q RoPE should modify its input in place or return a new tensor.

Views alone do not cause this issue. When a normal view of a differentiable non-leaf tensor is modified in place, autograd rebases the view’s history onto its base and records a CopySlices node. During backward, CopySlices reconstructs the gradient for the base tensor.

For our flow, this causes two problems
* MXFP8Linear returns a view created inside a custom autograd.Function. PyTorch forbids modifying it in place because view-mutation handling could conflict with its custom backward.
* After making MXFP8Linear’s output owned, Attention still creates views with view() and unsqueeze(). Mutating through these views inserts CopySlices, causing backward allocations and copies. These cannot be fused into the existing RoPE kernel.

Proposed Solutions
-------------------

### In-place

The main point for inplace is that **#4665 is not enough**. We still have copy slices due to attention view/unsqueeze ops, and this leads to suboptimal perf in the backward. 

The only advantage here is forward perf - approximately 14.1 µs for the measured Q RoPE kernel vs 60.5us. But there is so much complexity and there is so much coupling across mxfp8 linear, attention views and fused Rope kernel.

### out of place
This is simple, no CopySlice business. The only wrinkle is that we have to do one more memory read and this leads to slighly higher forward time. 

But overall, this is way simpler and composable. This is resilient to future model changes. In addition, for the training, the total forward/backward time is better for out-of-place - forward is slower but the backward is faster little bit more due to absence of copy slices

```

  In-place through Attention view:  5698 µs/iteration
  Out-of-place:                     5500 µs/iteration
```

**Review by anijain2305 — APPROVED** [review]

**Review by tianyu-l — APPROVED** [review]
